packages/core-react/src/hooks/useAutoConnect.ts

```diff
@@ -0,0 +1,102 @@
+import { useClient, useAccount, Chain } from 'wagmi';
+import { connect, disconnect } from 'wagmi/actions';
+import type { ConnectResult, Connector } from '@wagmi/core';
+import { useContext, useEffect, useRef } from 'react';
+import { WalletAdapterData } from '@reef-knot/types';
+import { AcceptTermsModalContext } from '../context/acceptTermsModal';
+import { LS_KEY_TERMS_ACCEPTANCE } from '../constants/localStorage';
+import { getUnsupportedChainError } from '../helpers/getUnsupportedChainError';
+
+const isTermsAccepted = () => {
+  if (typeof window !== 'undefined') {
+    return window.localStorage?.getItem(LS_KEY_TERMS_ACCEPTANCE) === 'true';
+  }
+  return false;
+};
+
+export const useAutoConnect = (
+  autoConnectEnabled: boolean,
+  walletDataList: WalletAdapterData[],
+  chains: Chain[],
+) => {
+  const isAutoConnectCalled = useRef(false);
+  const client = useClient();
+  const { isConnected } = useAccount();
+  const { acceptTermsModal } = useContext(AcceptTermsModalContext);
+  const autoConnectOnlyAdapters = walletDataList.filter(
+    (adapter) => adapter.autoConnectOnly,
+  );
+
+  useEffect(() => {
+    if (isConnected || !autoConnectEnabled || isAutoConnectCalled.current)
+      return;
+
+    void (async () => {
+      isAutoConnectCalled.current = true;
+      let connectResult: ConnectResult | null = null;
+      let connectError: Error | undefined;
+
+      const connectAndHandleErrors = async (connector: Connector) => {
+        connectError = undefined; // reset previous error if any
+        try {
+          connectResult = await connect({ connector });
+        } catch (e) {
+          connectResult = null; // ensure that connectResult is empty in case of an error
+          connectError = e as Error;
```

Comment:
It looks like if there is an error you need to reset Ref too

Reply:
Do you mean `isAutoConnectCalled` ref?
In this case, resetting `isAutoConnectCalled` ref can lead to the endless cycle:
`autoconnect => error => autoconnect => error => ...`
Current logic is to try autoconnect only once. If error, then show a modal with this error and suggest a "Connect" button.